File: 6-dangling-heap-pointers/code.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

// Represents a resource with an ID and data payload
typedef struct {
    int id;
    char* name;
    double* values;
    int values_count;
} Resource;

// Global cache of resource references (not owners!)
Resource** resource_cache;
int cache_size = 0;
int cache_capacity = 10;

// Initialize the cache
void init_cache() {
    resource_cache = (Resource**)malloc(sizeof(Resource*) * cache_capacity);
    for (int i = 0; i < cache_capacity; i++) {
        resource_cache[i] = NULL;
    }
}

// Add a resource reference to cache (does NOT take ownership)
void cache_resource(Resource* res) {
    if (cache_size < cache_capacity) {
        resource_cache[cache_size++] = res;
        printf("Resource %d (%s) cached at index %d\n", res->id, res->name, cache_size - 1);
    }
}

// Create a new resource
Resource* create_resource(int id, const char* name, int values_count) {
    Resource* res = (Resource*)malloc(sizeof(Resource));
    
    res->id = id;
    res->name = (char*)malloc(strlen(name) + 1);
    strcpy(res->name, name);
    
    res->values_count = values_count;
    res->values = (double*)malloc(sizeof(double) * values_count);
    
    // Initialize with some values
    for (int i = 0; i < values_count; i++) {
        res->values[i] = id * 100.0 + i;
    }
    
    printf("Created resource %d (%s) with %d values\n", id, name, values_count);
    return res;
}

// Free a resource's memory
void free_resource(Resource* res) {
    if (res != NULL) {
        printf("Freeing resource %d (%s)...\n", res->id, res->name);
        free(res->name);
        free(res->values);
        free(res);
    }
}

// Print resource details
void print_resource(Resource* res) {
    printf("Resource %d: %s\n", res->id, res->name);
    printf("Values: ");
    for (int i = 0; i < res->values_count; i++) {
        printf("%.1f ", res->values[i]);
    }
    printf("\n");
}

// Access cached resources (dangerous after resource is freed)
void access_cached_resources() {
    printf("\nAccessing cached resources:\n");
    for (int i = 0; i < cache_size; i++) {
        printf("Cache index %d: ", i);
        
        // Dangerous: dereferencing potentially freed pointers
        print_resource(resource_cache[i]); // Potential dangling heap pointer dereference!
    }
}

int main() {
    clock_t start, end;
    double cpu_time_used;
    
    start = clock();
    
    // Initialize our resource cache
    init_cache();
    
    // Create resources and add references to cache
    Resource* res1 = create_resource(1, "First Resource", 3);
    Resource* res2 = create_resource(2, "Second Resource", 5);
    Resource* res3 = create_resource(3, "Third Resource", 2);
    
    // Cache the resources (just references, not ownership)
    cache_resource(res1);
    cache_resource(res2);
    cache_resource(res3);
    
    printf("\nAccessing resources (safe at this point):\n");
    access_cached_resources();
    
    // Now free the first and third resources
    printf("\nFreeing resources 1 and 3...\n");
    free_resource(res1);
    free_resource(res3);
    
    // WARNING: res1 and res3 now point to freed memory!
    // But they're still in the cache...
    
    printf("\nAttempting to access cached resources after freeing 1 and 3:\n");
    // This will cause dangling heap pointer dereferences for res1 and res3
    access_cached_resources();
    
    // Clean up resources
    free_resource(res2); // Only free res2, as res1 and res3 are already freed
    free(resource_cache);
    
    end = clock();
    cpu_time_used = ((double) (end - start)) / CLOCKS_PER_SEC;
    printf("\nExecution time: %f seconds\n", cpu_time_used);
    
    return 0;
}

File: 6-dangling-heap-pointers/code.rs
use std::time::Instant;

// Represents a resource with an ID and data payload
struct Resource {
    id: i32,
    name: String,
    values: Vec<f64>,
}

// Implementation details for Resource
impl Resource {
    // Create a new resource
    fn new(id: i32, name: &str, values_count: usize) -> Self {
        let mut values = Vec::with_capacity(values_count);
        
        // Initialize with some values
        for i in 0..values_count {
            values.push(id as f64 * 100.0 + i as f64);
        }
        
        println!("Created resource {} ({}) with {} values", id, name, values_count);
        
        Resource {
            id,
            name: name.to_string(),
            values,
        }
    }
    
    // Print resource details
    fn print(&self) {
        println!("Resource {}: {}", self.id, self.name);
        print!("Values: ");
        for val in &self.values {
            print!("{:.1} ", val);
        }
        println!();
    }
}

// Global cache of resource references (raw pointers, not owners!)
struct ResourceCache {
    cache: Vec<*const Resource>,
}

impl ResourceCache {
    // Initialize the cache
    fn new(capacity: usize) -> Self {
        ResourceCache {
            cache: Vec::with_capacity(capacity),
        }
    }
    
    // Add a resource reference to cache (does NOT take ownership)
    fn cache_resource(&mut self, res: &Resource) {
        let ptr = res as *const Resource;
        self.cache.push(ptr);
        println!("Resource {} ({}) cached at index {}", res.id, res.name, self.cache.len() - 1);
    }
    
    // Access cached resources (dangerous after resource is dropped)
    unsafe fn access_cached_resources(&self) {
        println!("\nAccessing cached resources:");
        for (i, &ptr) in self.cache.iter().enumerate() {
            print!("Cache index {}: ", i);
            
            // Dangerous: dereferencing potentially dangling pointers
            let res_ref = &*ptr; // Potential dangling heap pointer dereference!
            res_ref.print();
        }
    }
}

fn main() {
    let start = Instant::now();
    
    // Initialize our resource cache
    let mut cache = ResourceCache::new(10);
    
    // Create resources
    let res1 = Resource::new(1, "First Resource", 3);
    let res2 = Resource::new(2, "Second Resource", 5);
    let res3 = Resource::new(3, "Third Resource", 2);
    
    // Cache the resources (just references, not ownership)
    cache.cache_resource(&res1);
    cache.cache_resource(&res2);
    cache.cache_resource(&res3);
    
    println!("\nAccessing resources (safe at this point):");
    unsafe {
        cache.access_cached_resources();
    }
    
    // Now let's create a scope where res1 and res3 will be dropped at the end
    {
        // Move res1 and res3 into this scope
        let _temp_res1 = res1;
        let _temp_res3 = res3;
        
        println!("\nResources 1 and 3 will be dropped at the end of this scope...");
    } // res1 and res3 are dropped here!
    
    // WARNING: cache still has pointers to res1 and res3, but they are now dangling!
    
    println!("\nAttempting to access cached resources after dropping 1 and 3:");
    // This will cause dangling heap pointer dereferences for res1 and res3
    unsafe {
        cache.access_cached_resources();
    }
    
    let duration = start.elapsed();
    println!("\nExecution time: {:?}", duration);
}